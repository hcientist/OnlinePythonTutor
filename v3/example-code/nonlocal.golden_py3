{
  "code": "# 'nonlocal' keyword is only in Python 3\ndef outer():\n    x = 1\n    def inner():\n        nonlocal x\n        x = 2\n        y = x\n        print(\"inner:\", x, y)\n    inner()\n    print(\"outer:\", x)\n\nouter()\n", 
  "trace": [
    {
      "ordered_globals": [], 
      "stdout": "", 
      "func_name": "<module>", 
      "stack_to_render": [], 
      "globals": {}, 
      "heap": {}, 
      "line": 2, 
      "event": "step_line"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "", 
      "func_name": "<module>", 
      "stack_to_render": [], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ]
      }, 
      "line": 12, 
      "event": "step_line"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "", 
      "func_name": "outer", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {}, 
          "is_highlighted": true, 
          "is_parent": false, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1", 
          "ordered_varnames": []
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ]
      }, 
      "line": 2, 
      "event": "call"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "", 
      "func_name": "outer", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {}, 
          "is_highlighted": true, 
          "is_parent": false, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1", 
          "ordered_varnames": []
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ]
      }, 
      "line": 3, 
      "event": "step_line"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "", 
      "func_name": "outer", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {
            "x": 1
          }, 
          "is_highlighted": true, 
          "is_parent": false, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1", 
          "ordered_varnames": [
            "x"
          ]
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ]
      }, 
      "line": 4, 
      "event": "step_line"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "", 
      "func_name": "outer", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {
            "x": 1, 
            "inner": [
              "REF", 
              2
            ]
          }, 
          "is_highlighted": true, 
          "is_parent": true, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1_p", 
          "ordered_varnames": [
            "inner", 
            "x"
          ]
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ], 
        "2": [
          "FUNCTION", 
          "inner()", 
          1
        ]
      }, 
      "line": 9, 
      "event": "step_line"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "", 
      "func_name": "inner", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {
            "x": 1, 
            "inner": [
              "REF", 
              2
            ]
          }, 
          "is_highlighted": false, 
          "is_parent": true, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1_p", 
          "ordered_varnames": [
            "inner", 
            "x"
          ]
        }, 
        {
          "frame_id": 2, 
          "encoded_locals": {}, 
          "is_highlighted": true, 
          "is_parent": false, 
          "func_name": "inner", 
          "is_zombie": false, 
          "parent_frame_id_list": [
            1
          ], 
          "unique_hash": "inner_f2", 
          "ordered_varnames": []
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ], 
        "2": [
          "FUNCTION", 
          "inner()", 
          1
        ]
      }, 
      "line": 4, 
      "event": "call"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "", 
      "func_name": "inner", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {
            "x": 1, 
            "inner": [
              "REF", 
              2
            ]
          }, 
          "is_highlighted": false, 
          "is_parent": true, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1_p", 
          "ordered_varnames": [
            "inner", 
            "x"
          ]
        }, 
        {
          "frame_id": 2, 
          "encoded_locals": {}, 
          "is_highlighted": true, 
          "is_parent": false, 
          "func_name": "inner", 
          "is_zombie": false, 
          "parent_frame_id_list": [
            1
          ], 
          "unique_hash": "inner_f2", 
          "ordered_varnames": []
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ], 
        "2": [
          "FUNCTION", 
          "inner()", 
          1
        ]
      }, 
      "line": 6, 
      "event": "step_line"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "", 
      "func_name": "inner", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {
            "x": 2, 
            "inner": [
              "REF", 
              2
            ]
          }, 
          "is_highlighted": false, 
          "is_parent": true, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1_p", 
          "ordered_varnames": [
            "inner", 
            "x"
          ]
        }, 
        {
          "frame_id": 2, 
          "encoded_locals": {}, 
          "is_highlighted": true, 
          "is_parent": false, 
          "func_name": "inner", 
          "is_zombie": false, 
          "parent_frame_id_list": [
            1
          ], 
          "unique_hash": "inner_f2", 
          "ordered_varnames": []
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ], 
        "2": [
          "FUNCTION", 
          "inner()", 
          1
        ]
      }, 
      "line": 7, 
      "event": "step_line"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "", 
      "func_name": "inner", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {
            "x": 2, 
            "inner": [
              "REF", 
              2
            ]
          }, 
          "is_highlighted": false, 
          "is_parent": true, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1_p", 
          "ordered_varnames": [
            "inner", 
            "x"
          ]
        }, 
        {
          "frame_id": 2, 
          "encoded_locals": {
            "y": 2
          }, 
          "is_highlighted": true, 
          "is_parent": false, 
          "func_name": "inner", 
          "is_zombie": false, 
          "parent_frame_id_list": [
            1
          ], 
          "unique_hash": "inner_f2", 
          "ordered_varnames": [
            "y"
          ]
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ], 
        "2": [
          "FUNCTION", 
          "inner()", 
          1
        ]
      }, 
      "line": 8, 
      "event": "step_line"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "inner: 2 2\n", 
      "func_name": "inner", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {
            "x": 2, 
            "inner": [
              "REF", 
              2
            ]
          }, 
          "is_highlighted": false, 
          "is_parent": true, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1_p", 
          "ordered_varnames": [
            "inner", 
            "x"
          ]
        }, 
        {
          "frame_id": 2, 
          "encoded_locals": {
            "y": 2, 
            "__return__": null
          }, 
          "is_highlighted": true, 
          "is_parent": false, 
          "func_name": "inner", 
          "is_zombie": false, 
          "parent_frame_id_list": [
            1
          ], 
          "unique_hash": "inner_f2", 
          "ordered_varnames": [
            "y", 
            "__return__"
          ]
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ], 
        "2": [
          "FUNCTION", 
          "inner()", 
          1
        ]
      }, 
      "line": 8, 
      "event": "return"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "inner: 2 2\n", 
      "func_name": "outer", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {
            "x": 2, 
            "inner": [
              "REF", 
              2
            ]
          }, 
          "is_highlighted": true, 
          "is_parent": true, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1_p", 
          "ordered_varnames": [
            "inner", 
            "x"
          ]
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ], 
        "2": [
          "FUNCTION", 
          "inner()", 
          1
        ]
      }, 
      "line": 10, 
      "event": "step_line"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "inner: 2 2\nouter: 2\n", 
      "func_name": "outer", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {
            "__return__": null, 
            "x": 2, 
            "inner": [
              "REF", 
              2
            ]
          }, 
          "is_highlighted": true, 
          "is_parent": true, 
          "func_name": "outer", 
          "is_zombie": false, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1_p", 
          "ordered_varnames": [
            "inner", 
            "x", 
            "__return__"
          ]
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ], 
        "2": [
          "FUNCTION", 
          "inner()", 
          1
        ]
      }, 
      "line": 10, 
      "event": "return"
    }, 
    {
      "ordered_globals": [
        "outer"
      ], 
      "stdout": "inner: 2 2\nouter: 2\n", 
      "func_name": "<module>", 
      "stack_to_render": [
        {
          "frame_id": 1, 
          "encoded_locals": {
            "__return__": null, 
            "x": 2, 
            "inner": [
              "REF", 
              2
            ]
          }, 
          "is_highlighted": false, 
          "is_parent": true, 
          "func_name": "outer", 
          "is_zombie": true, 
          "parent_frame_id_list": [], 
          "unique_hash": "outer_f1_p_z", 
          "ordered_varnames": [
            "inner", 
            "x", 
            "__return__"
          ]
        }
      ], 
      "globals": {
        "outer": [
          "REF", 
          1
        ]
      }, 
      "heap": {
        "1": [
          "FUNCTION", 
          "outer()", 
          null
        ], 
        "2": [
          "FUNCTION", 
          "inner()", 
          1
        ]
      }, 
      "line": 12, 
      "event": "return"
    }
  ]
}
