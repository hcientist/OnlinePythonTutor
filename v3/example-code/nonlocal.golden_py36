{
  "code": "# 'nonlocal' keyword is only in Python 3\ndef outer():\n    x = 1\n    def inner():\n        nonlocal x\n        x = 2\n        y = x\n        print(\"inner:\", x, y)\n    inner()\n    print(\"outer:\", x)\n\nouter()\n",
  "trace": [
    {
      "line": 2,
      "event": "step_line",
      "func_name": "<module>",
      "globals": {},
      "ordered_globals": [],
      "stack_to_render": [],
      "heap": {},
      "stdout": ""
    },
    {
      "line": 12,
      "event": "step_line",
      "func_name": "<module>",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [],
      "heap": {
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": ""
    },
    {
      "line": 2,
      "event": "call",
      "func_name": "outer",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": false,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {},
          "ordered_varnames": [],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "outer_f1"
        }
      ],
      "heap": {
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": ""
    },
    {
      "line": 3,
      "event": "step_line",
      "func_name": "outer",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": false,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {},
          "ordered_varnames": [],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "outer_f1"
        }
      ],
      "heap": {
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": ""
    },
    {
      "line": 4,
      "event": "step_line",
      "func_name": "outer",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": false,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {
            "x": 1
          },
          "ordered_varnames": [
            "x"
          ],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "outer_f1"
        }
      ],
      "heap": {
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": ""
    },
    {
      "line": 9,
      "event": "step_line",
      "func_name": "outer",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": true,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {
            "x": 1,
            "inner": [
              "REF",
              2
            ]
          },
          "ordered_varnames": [
            "inner",
            "x"
          ],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "outer_f1_p"
        }
      ],
      "heap": {
        "2": [
          "FUNCTION",
          "inner()",
          1
        ],
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": ""
    },
    {
      "line": 4,
      "event": "call",
      "func_name": "inner",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": true,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {
            "x": 1,
            "inner": [
              "REF",
              2
            ]
          },
          "ordered_varnames": [
            "inner",
            "x"
          ],
          "is_zombie": false,
          "is_highlighted": false,
          "unique_hash": "outer_f1_p"
        },
        {
          "func_name": "inner",
          "is_parent": false,
          "frame_id": 2,
          "parent_frame_id_list": [
            1
          ],
          "encoded_locals": {},
          "ordered_varnames": [],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "inner_f2"
        }
      ],
      "heap": {
        "2": [
          "FUNCTION",
          "inner()",
          1
        ],
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": ""
    },
    {
      "line": 6,
      "event": "step_line",
      "func_name": "inner",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": true,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {
            "x": 1,
            "inner": [
              "REF",
              2
            ]
          },
          "ordered_varnames": [
            "inner",
            "x"
          ],
          "is_zombie": false,
          "is_highlighted": false,
          "unique_hash": "outer_f1_p"
        },
        {
          "func_name": "inner",
          "is_parent": false,
          "frame_id": 2,
          "parent_frame_id_list": [
            1
          ],
          "encoded_locals": {},
          "ordered_varnames": [],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "inner_f2"
        }
      ],
      "heap": {
        "2": [
          "FUNCTION",
          "inner()",
          1
        ],
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": ""
    },
    {
      "line": 7,
      "event": "step_line",
      "func_name": "inner",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": true,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {
            "x": 2,
            "inner": [
              "REF",
              2
            ]
          },
          "ordered_varnames": [
            "inner",
            "x"
          ],
          "is_zombie": false,
          "is_highlighted": false,
          "unique_hash": "outer_f1_p"
        },
        {
          "func_name": "inner",
          "is_parent": false,
          "frame_id": 2,
          "parent_frame_id_list": [
            1
          ],
          "encoded_locals": {},
          "ordered_varnames": [],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "inner_f2"
        }
      ],
      "heap": {
        "2": [
          "FUNCTION",
          "inner()",
          1
        ],
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": ""
    },
    {
      "line": 8,
      "event": "step_line",
      "func_name": "inner",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": true,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {
            "x": 2,
            "inner": [
              "REF",
              2
            ]
          },
          "ordered_varnames": [
            "inner",
            "x"
          ],
          "is_zombie": false,
          "is_highlighted": false,
          "unique_hash": "outer_f1_p"
        },
        {
          "func_name": "inner",
          "is_parent": false,
          "frame_id": 2,
          "parent_frame_id_list": [
            1
          ],
          "encoded_locals": {
            "y": 2
          },
          "ordered_varnames": [
            "y"
          ],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "inner_f2"
        }
      ],
      "heap": {
        "2": [
          "FUNCTION",
          "inner()",
          1
        ],
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": ""
    },
    {
      "line": 8,
      "event": "return",
      "func_name": "inner",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": true,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {
            "x": 2,
            "inner": [
              "REF",
              2
            ]
          },
          "ordered_varnames": [
            "inner",
            "x"
          ],
          "is_zombie": false,
          "is_highlighted": false,
          "unique_hash": "outer_f1_p"
        },
        {
          "func_name": "inner",
          "is_parent": false,
          "frame_id": 2,
          "parent_frame_id_list": [
            1
          ],
          "encoded_locals": {
            "y": 2,
            "__return__": null
          },
          "ordered_varnames": [
            "y",
            "__return__"
          ],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "inner_f2"
        }
      ],
      "heap": {
        "2": [
          "FUNCTION",
          "inner()",
          1
        ],
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": "inner: 2 2\n"
    },
    {
      "line": 10,
      "event": "step_line",
      "func_name": "outer",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": true,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {
            "x": 2,
            "inner": [
              "REF",
              2
            ]
          },
          "ordered_varnames": [
            "inner",
            "x"
          ],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "outer_f1_p"
        }
      ],
      "heap": {
        "2": [
          "FUNCTION",
          "inner()",
          1
        ],
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": "inner: 2 2\n"
    },
    {
      "line": 10,
      "event": "return",
      "func_name": "outer",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": true,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {
            "x": 2,
            "inner": [
              "REF",
              2
            ],
            "__return__": null
          },
          "ordered_varnames": [
            "inner",
            "x",
            "__return__"
          ],
          "is_zombie": false,
          "is_highlighted": true,
          "unique_hash": "outer_f1_p"
        }
      ],
      "heap": {
        "2": [
          "FUNCTION",
          "inner()",
          1
        ],
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": "inner: 2 2\nouter: 2\n"
    },
    {
      "line": 12,
      "event": "return",
      "func_name": "<module>",
      "globals": {
        "outer": [
          "REF",
          1
        ]
      },
      "ordered_globals": [
        "outer"
      ],
      "stack_to_render": [
        {
          "func_name": "outer",
          "is_parent": true,
          "frame_id": 1,
          "parent_frame_id_list": [],
          "encoded_locals": {
            "x": 2,
            "inner": [
              "REF",
              2
            ],
            "__return__": null
          },
          "ordered_varnames": [
            "inner",
            "x",
            "__return__"
          ],
          "is_zombie": true,
          "is_highlighted": false,
          "unique_hash": "outer_f1_p_z"
        }
      ],
      "heap": {
        "2": [
          "FUNCTION",
          "inner()",
          1
        ],
        "1": [
          "FUNCTION",
          "outer()",
          null
        ]
      },
      "stdout": "inner: 2 2\nouter: 2\n"
    }
  ]
}
