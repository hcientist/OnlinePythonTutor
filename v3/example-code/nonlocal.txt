# 'nonlocal' keyword is only in Python 3
def outer():
    x = 1
    def inner():
        nonlocal x
        x = 2
        y = x
        print("inner:", x, y)
    inner()
    print("outer:", x)

outer()
