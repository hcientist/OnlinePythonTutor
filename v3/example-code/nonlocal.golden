{
  "code": "# 'nonlocal' keyword is only in Python 3\ndef outer():\n    x = 1\n    def inner():\n        nonlocal x\n        x = 2\n        y = x\n        print(\"inner:\", x, y)\n    inner()\n    print(\"outer:\", x)\n\nouter()\n", 
  "trace": [
    {
      "exception_msg": "SyntaxError: invalid syntax (<string>, line 5)", 
      "line": 5, 
      "event": "uncaught_exception", 
      "offset": 18
    }
  ]
}
